Sources/Instrumentation/MetricKit/MetricKitInstrumentation.swift

```diff
@@ -0,0 +1,608 @@
+#if canImport(MetricKit) && !os(tvOS)
+    import Foundation
+    import MetricKit
+    import OpenTelemetryApi
+
+    private let metricKitInstrumentationName = "MetricKit"
+    private let metricKitInstrumentationVersion = "0.0.1"
+
+    @available(iOS 13.0, macOS 12.0, macCatalyst 13.1, visionOS 1.0, *)
+    public class MetricKitInstrumentation: NSObject, MXMetricManagerSubscriber {
+        public func didReceive(_ payloads: [MXMetricPayload]) {
+            for payload in payloads {
+                reportMetrics(payload: payload)
+            }
+        }
+
+        @available(iOS 14.0, macOS 12.0, macCatalyst 14.0, watchOS 7.0, *)
+        public func didReceive(_ payloads: [MXDiagnosticPayload]) {
+            for payload in payloads {
+                reportDiagnostics(payload: payload)
+            }
+        }
+    }
+
+    // MARK: - MetricKit helpers
+
+    func getMetricKitTracer() -> Tracer {
+        return OpenTelemetry.instance.tracerProvider.get(
+            instrumentationName: metricKitInstrumentationName,
+            instrumentationVersion: metricKitInstrumentationVersion,
+        )
+    }
+
+    /// Estimates the average value of the whole histogram.
+    @available(iOS 13.0, macOS 12.0, macCatalyst 13.1, visionOS 1.0, *)
+    func estimateHistogramAverage<UnitType>(_ histogram: MXHistogram<UnitType>) -> Measurement<
+        UnitType
+    >? {
+        var estimatedSum: Measurement<UnitType>?
+        var sampleCount = 0.0
+        for bucket in histogram.bucketEnumerator {
+            let bucket = bucket as! MXHistogramBucket<UnitType>
+            let estimatedValue = (bucket.bucketStart + bucket.bucketEnd) / 2.0
+            let count = Double(bucket.bucketCount)
+            estimatedSum =
+                if let previousSum = estimatedSum {
+                    previousSum + estimatedValue * count
+                } else {
+                    estimatedValue * count
+                }
+            sampleCount += count
+        }
+        return estimatedSum.map { $0 / sampleCount }
+    }
+
+    @available(iOS 13.0, macOS 12.0, macCatalyst 13.1, visionOS 1.0, *)
+    public func reportMetrics(payload: MXMetricPayload) {
+        let span = getMetricKitTracer().spanBuilder(spanName: "MXMetricPayload")
```

Comment:
I think there was some debate about capturing `MXMetric` as a metric, span, or log. Could you document your decision to record this as a span? 

With the way its setup now, seems that we can also extend this to support metrics.